compiler/noirc_frontend/src/hir/def_collector/dc_mod.rs

```diff
@@ -316,6 +316,85 @@ impl<'a> ModCollector<'a> {
         errors
     }
 
+    fn check_for_duplicate_trait_items(
```

Comment:
You shouldn't need an entire new function to check for this. Let's try to handle it the same way we do for functions (or struct methods) by inserting the name into a module. We'll need to do this for `Trait::method(args)` syntax anyway. See e.g. dc_mod.rs line 219 for where function duplicates are checked.

Reply:
What is this `Trait::method(args)` syntax that you're talking about? You certainly can't call a trait default method, can you? A trait by itself is incomplete. A default implementation of a trait method can e.g. call other methods, that don't have a default implementation, and it can use types and consts that don't have a default value as well. So, you certainly can't call it. Although, I agree that at some point we still need to collect it and type check it, in order to catch errors in traits, that don't have any implementations.

Reply:
Let me clarify this: You cannot call a method via the trait namespace like this `Trait::method(args)`. You can call it in one of the following ways:
1.`Self::method(args)` from another function inside the trait
2.`T::method(args)` from a function with a trait bound `where T: Trait`
3.`Struct::method(args)` for a struct that has a trait impl
Why should we add the functions to the trait namespace, then?

Reply:
The main reason I suggested putting them in a trait namespace now though was so that we do not need to write a new function to check for these error cases. They should already be covered by the existing functions to add an item to a module.

Reply:
Hmm, why doesn't this work in Rust, then?

    trait MyTrait {
        fn MyAdd(a: u32, b:u32) -> u32;
    }
    
    impl MyTrait for u32 {
        fn MyAdd(a: u32, b:u32) -> u32 {
            a + b
        }
    }
    
    fn main() {
        let x: u32 = Default::default();
        let r = MyTrait::MyAdd(1, 2);
        println!("zero = {x}, three = {r}");
    }
I get this error:

    error[E0790]: cannot call associated function on trait without specifying the corresponding `impl` type

What's the difference between this and your example, which calls std::ops::Add::add()?

Reply:
@nickysn ah, your trait definition has no usage of `self` or `Self` anywhere, so the type is ambiguous. Since your trait was defined to always take two `u32`s and return a `u32` regardless of which type it is impld for, the type system cannot figure out which impl to use unless you explicitly specify it. Hence the error message.

In my example, `Default::default` returns a `Self` type, so when I said `let x: u32 = Default::default()` it knew to call `<u32 as Default>::default()`. Likewise with `Add`, `add` takes in a `self` argument as well as an `other: Self`. When  I provided two integers (which default to i32), it was able to figure out I meant `<i32 as std::ops::Add>::add(1, 2)`.

Reply:
So, Rust uses type inference to determine the type for the trait, based on the function parameters and the presence of a self parameter. Ouch! Sounds like more things for the traits roadmap...

Reply:
It should hopefully be a similar mechanism to normal trait method resolution. You could e.g. transform each `a.trait_method(b)` call to `Trait::trait_method(a, b)`. Then you'd only have the one case to handle afterward.

Reply:
Ok, refactored the duplicate check, so that it uses the module system. This even exposed an unrelated bug that we already had, but was not caught previously - trait methods were registered in the wrong crate, when the struct is in a different crate. Registering the trait methods caused the traits_in_crates_1 test to fail by a spurious duplicate method error. So, this bug was fixed as well.